x-pack/plugins/ecs_data_quality_dashboard/server/plugin.ts

```diff
@@ -5,34 +5,83 @@
  * 2.0.
  */
 
-import { PluginInitializerContext, CoreSetup, CoreStart, Plugin, Logger } from '@kbn/core/server';
+import type {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+} from '@kbn/core/server';
 
-import { EcsDataQualityDashboardPluginSetup, EcsDataQualityDashboardPluginStart } from './types';
+import { ReplaySubject, type Subject } from 'rxjs';
+import type { DataStreamSpacesAdapter } from '@kbn/data-stream-adapter';
+import { DEFAULT_SPACE_ID } from '@kbn/spaces-plugin/common/constants';
+import type {
+  EcsDataQualityDashboardPluginSetup,
+  EcsDataQualityDashboardPluginStart,
+  PluginSetupDependencies,
+  DataQualityDashboardRequestHandlerContext,
+} from './types';
 import {
   getILMExplainRoute,
   getIndexMappingsRoute,
   getIndexStatsRoute,
   getUnallowedFieldValuesRoute,
+  resultsRoutes,
 } from './routes';
+import { createResultsDataStream } from './lib/data_stream/results_data_stream';
 
 export class EcsDataQualityDashboardPlugin
   implements Plugin<EcsDataQualityDashboardPluginSetup, EcsDataQualityDashboardPluginStart>
 {
   private readonly logger: Logger;
+  private readonly resultsDataStream: DataStreamSpacesAdapter;
+  private pluginStop$: Subject<void>;
 
   constructor(initializerContext: PluginInitializerContext) {
     this.logger = initializerContext.logger.get();
+    this.pluginStop$ = new ReplaySubject(1);
+    this.resultsDataStream = createResultsDataStream({
+      kibanaVersion: initializerContext.env.packageInfo.version,
+    });
   }
 
-  public setup(core: CoreSetup) {
-    this.logger.debug('ecsDataQualityDashboard: Setup'); // this would be deleted when plugin is removed
-    const router = core.http.createRouter(); // this would be deleted when plugin is removed
+  public setup(core: CoreSetup, plugins: PluginSetupDependencies) {
+    this.logger.debug('ecsDataQualityDashboard: Setup');
+
+    core.getStartServices().then(([{ elasticsearch }]) => {
+      this.resultsDataStream.install({
+        esClient: elasticsearch.client.asInternalUser,
+        logger: this.logger,
+        pluginStop$: this.pluginStop$,
+      });
+    });
+
+    core.http.registerRouteHandlerContext<
+      DataQualityDashboardRequestHandlerContext,
+      'dataQualityDashboard'
+    >('dataQualityDashboard', (_context, request) => {
+      const spaceId = plugins.spaces?.spacesService.getSpaceId(request) ?? DEFAULT_SPACE_ID;
+      return {
+        spaceId,
+        getResultsIndexName: async (): Promise<string> => {
+          let indexName = await this.resultsDataStream.getSpaceIndexName(spaceId);
+          if (!indexName) {
+            indexName = await this.resultsDataStream.installSpace(spaceId);
```

Comment:
**Summary**

Consider whether or not Kibana should crash at startup if `this.resultsDataStream.install()` fails when called in the `EcsDataQualityDashboardPlugin`'s `setup()`

**Details**

It appears (based on an experiment detailed below), that Kibana will crash at startup if the data stream for storing Data Quality (DQ) Dashboard results cannot be installed when `this.resultsDataStream.install()` is called in the `EcsDataQualityDashboardPlugin` `setup()`.

1) Given that the current implementation of the Data Quality dashboard works correctly without persistence of results, should startup of Kibana continue (instead of crashing) if the data stream for persisting DQ results cannot be installed?

2) If the answer to 1 is "yes, Kibana should start without crashing", should the `kbn-data-stream-adapter` provide an API response that makes troubleshooting data stream creation possible with little or no context from log files?

**Reproduction steps**

To simulate failure:

1) Add the following `console.log` to `x-pack/plugins/ecs_data_quality_dashboard/server/plugin.ts`

```ts
  public setup(core: CoreSetup, plugins: PluginSetupDependencies) {
    this.logger.debug('ecsDataQualityDashboard: Setup');

    core.getStartServices().then(([{ elasticsearch }]) => {
      console.log('--> before resultsDataStream.install');
      this.resultsDataStream.install({
        esClient: elasticsearch.client.asInternalUser,
        logger: this.logger,
        pluginStop$: this.pluginStop$,
      });
    });
// ...
```

2) Edit the `createOrUpdateComponentTemplateHelper` function in `packages/kbn-data-stream-adapter/src/create_or_update_component_template.ts` such that it throws the simulated error below:

```ts
const createOrUpdateComponentTemplateHelper = async (
  esClient: ElasticsearchClient,
  template: ClusterPutComponentTemplateRequest,
  totalFieldsLimit: number,
  logger: Logger
) => {
  try {
    console.error('--> simulated error in createOrUpdateComponentTemplateHelper');
    throw new Error('simulated error in createOrUpdateComponentTemplateHelper');

    // await retryTransientEsErrors(() => esClient.cluster.putComponentTemplate(template), { logger });
  } catch (error) {
```

3) Start Kibana via `yarn start --no-base-path`

Kibana will crash at startup per the following example output:

```
--> before resultsDataStream.install
[2024-01-12T23:06:16.630-07:00][INFO ][plugins.ecsDataQualityDashboard] Installing component template .kibana-dqa-dashboard-ecs-mappings
--> simulated error in createOrUpdateComponentTemplateHelper
[2024-01-12T23:06:16.630-07:00][INFO ][plugins.ecsDataQualityDashboard] Installing component template .kibana-dqa-dashboard-results-mappings
--> simulated error in createOrUpdateComponentTemplateHelper
[2024-01-12T23:06:16.638-07:00][INFO ][plugins.observability] Installing SLO shared resources
[2024-01-12T23:06:16.638-07:00][INFO ][plugins.observability] Installing SLO component template [.slo-observability.sli-mappings]
[2024-01-12T23:06:16.638-07:00][INFO ][plugins.observability] Installing SLO component template [.slo-observability.sli-settings]
[2024-01-12T23:06:16.639-07:00][INFO ][plugins.observability] Installing SLO component template [.slo-observability.summary-mappings]
[2024-01-12T23:06:16.639-07:00][INFO ][plugins.observability] Installing SLO component template [.slo-observability.summary-settings]
[2024-01-12T23:06:16.642-07:00][ERROR][plugins.ecsDataQualityDashboard] Error installing component template .kibana-dqa-dashboard-ecs-mappings - simulated error in createOrUpdateComponentTemplateHelper
[2024-01-12T23:06:16.642-07:00][ERROR][plugins.ecsDataQualityDashboard] Error installing component template .kibana-dqa-dashboard-results-mappings - simulated error in createOrUpdateComponentTemplateHelper
Unhandled Promise rejection detected:

Error: Failure during installation of create or update .kibana-dqa-dashboard-ecs-mappings component template. simulated error in createOrUpdateComponentTemplateHelper
    at installWithTimeout (install_with_timeout.ts:62:13)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at data_stream_adapter.ts:93:9
    at async Promise.all (index 0)
    at DataStreamSpacesAdapter.install (data_stream_spaces_adapter.ts:35:5)

Terminating process...
 server crashed  with status code 1
```

Reply:
Thanks for bringing up this point, @andrew-goldstein .

The observed behavior during DQ data stream installation errors is expected. I applied the same criteria as the 'Alerts-as-data' data stream creation. In case of any error (excluding manual 'shut down' errors) during installation the DQ plugin throws an error that pops up to the Kibana server, leading to the process stop.

While you're right that the DQ page can work without needing persistence, supporting DQ without persistence requires consideration across different parts of the application. We'd need to check for the data stream's existence and act accordingly.

Regarding the response to question 1, I'm curious to hear @dhru42's thoughts. In my opinion, the answer should be 'no':

I think we should only add a data stream existence check if we have an expected scenario in which DQ should work without persistence, like making it a gold license feature, for example. 
However, turning an unexpected scenario (data stream installation error) into an expected one in DQ logic might compromise the clarity of error detection for engineers, and create a user perception that errors during data stream creation are normal. 
Additionally, if we decide to support it, we should be consistent and consider applying the same criteria to every persistence-layer data stream we create, such as the AI Assistant.

While I don't dismiss the idea, I'm cautious about adding complexity to accommodate a scenario that should never happen. I would prefer maintaining a 'hard fail' response to installation errors. The data stream installation logic is now centralized in the new adapter library, errors should be rare unless there's a real problem.

What do you think?

Reply:
Thanks for bringing up this point Andrew,

The observed behavior during DQ data stream installation errors is expected. In case of any error (excluding manual 'shut down' errors) during data stream installation, it throws an error that pops up to the Kibana server, leading to a process stop. 

That was intentional, however, you are right that the DQ page works without persistence, the data stream is not strictly necessary. Nevertheless supporting DQ without persistence requires checks for the data stream's installation status in the API, I will have to address that as well.

Honestly, I am not very comfortable adding code to support unexpected scenarios like installation errors, but in this case, it probably makes sense to not crash the server, let it start, and do the check in the DQ results API, it will make the Security application more resilient. 

I am making the changes.